libraries/PDM/src/rp2040/PDM.cpp

```diff
@@ -118,14 +138,30 @@ int PDMClass::begin(int channels, int sampleRate) {
                           true                // Start immediately
                          );
 
+    _cutSamples = 100;
+
     _init = 1;
 
     return 1;
 }
 
 void PDMClass::end() {
+
+    if (_init != 1) {
+        return;
+    }
+
+    dma_channel_set_irq0_enabled(_dmaChannel, false);
     dma_channel_abort(_dmaChannel);
+    dma_channel_unclaim(_dmaChannel);
+    irq_remove_handler(DMA_IRQ_0, dmaHandler);
+    pio_remove_program(_pio,  &pdm_pio_program, _pgmOffset);
```

Comment:
You don't want to do this.  The way `PIOProgram` works is that it cached the uploaded address of a program so that multiple instances can be run without needing to upload multiple programs (i.e. you can do 4 SWSerial ports with only 1 pio-program in memory).  Leave the PIO program in memory and then the next time you begin the `PIOProgram` call will return the already-installed address